src/python/pants/subsystem/subsystem_client_mixin.py

```diff
@@ -53,8 +59,52 @@ def subsystem_dependencies(cls):
 
   @classmethod
   def subsystem_dependencies_iter(cls):
+    """Iterate over the direct subsystem dependencies of this Optionable."""
     for dep in cls.subsystem_dependencies():
       if isinstance(dep, SubsystemDependency):
         yield dep
       else:
         yield SubsystemDependency(dep, GLOBAL_SCOPE)
+
+  class CycleException(Exception):
+    """Thrown when a circular subsystem dependency is detected."""
+
+    def __init__(self, cycle):
+      message = 'Cycle detected:\n\t{}'.format(' ->\n\t'.join(
+        '{} scope: {}'.format(optionable_cls, optionable_cls.options_scope)
+        for optionable_cls in cycle))
+      super(SubsystemClientMixin.CycleException, self).__init__(message)
+
+  @classmethod
+  def known_scope_infos(cls):
+    """Yields ScopeInfo for all known scopes for this optionable, in no particular order.
+
+    :raises: :class:`pants.subsystem.subsystem_client_mixin.SubsystemClientMixin.CycleException`
+             if a dependency cycle is detected.
+    """
+    known_scope_infos = set()
+    optionables_path = OrderedSet()  #  To check for cycles at the Optionable level, ignoring scope.
+
+    def collect_scope_infos(optionable_cls, scoped_to):
```

Comment:
With the call `collect_scope_infos(cls, GLOBAL_SCOPE)` at the bottom, it looks like we are assuming the `cls` argument is necessarily an `Optionable` as well, which it doesn't seem is guaranteed. From a quick grep, it seems that every consumer of `SubsystemClientMixin` is also an `Optionable`, so it won't cause an error, but at that point is there a reason the contents of this mixin can't just be merged into Optionable? That's the approach I took in a very old pr that had similar motivations and I can't immediately see why it wouldn't apply here.

Reply:
There's a comment on `SubsystemClientMixin`: `Must be mixed in to an Optionable.`

The mixin isn't merged into `Optionable` because not every `Optionable` is a `SubsystemClientMixin`, although I suppose they could be. More to the point, the subsystem package depends on the option package, so that move would require merging the packages themselves, which may not be what we want. Currently `Optionable` is blissfully unaware of the existence of subsystems. Although granted, the subsystem package is tiny, and may not deserve independent existence. 

Still, my point is, that's for a future change, this one is already substantial enough.

Reply:
Ooh, I _definitely_ should have seen that comment -- I just asked because I wasn't sure if it had occurred to you. I definitely didn't realize the package dependency situation either. Thanks for the thorough explanation!